app/service/planning_svc.py

```diff
@@ -104,6 +107,34 @@ def __init__(self):
                 link['command'] = self.encode_string(decoded_test)
         return links
 
+    async def _do_enforcements(self, ability_requirements, operation, link, combo):
+        for requirements_info in ability_requirements:
+            # list of facts that have a relationship
+            requirements_info['fact_relationships'] = [fact for fact in operation['facts'] if
+                                                       len(fact['relationships']) > 0]
+            # list of used facts
+            uf = link.get('used', [])
+            requirement = await self._load_requirements(requirements_info)
+            if not requirement.enforce(combo[0], uf, operation['facts']):
+                return False
+        return True
+
+    @staticmethod
+    async def _get_operation_requirements(link, operation):
+        requirements = []
+        # get the requirement module from the operation dict
+        for phase in (operation['adversary']['phases']).values():
+            for ab in phase:
+                if ab['id'] == link['ability']:
+                    for req in ab.get('requirements', []):
+                        requirements.append(req)
+        return requirements
+
+    @staticmethod
+    async def _load_requirements(requirements_info):
```

Comment:
think you could extract this function to base_service? then this + the parsing (from parsing_svc) + the planner (from operation_svc) could use it.

Reply:
Sure, I think that's doable